src/OpenTelemetry/Trace/AWSXRayActivityTraceIdGenerator.cs

```diff
@@ -0,0 +1,123 @@
+
+        private const long TicksPerMicrosecond = TimeSpan.TicksPerMillisecond / 1000;
+        private const long MicrosecondPerSecond = TimeSpan.TicksPerSecond / TicksPerMicrosecond;
+
+        private static readonly DateTime EpochStart = new DateTime(1970, 1, 1, 0, 0, 0, DateTimeKind.Utc);
+        private static readonly long UnixEpochMicroseconds = EpochStart.Ticks / TicksPerMicrosecond;
+        private static readonly Random Global = new Random();
+        private static readonly ThreadLocal<Random> Local = new ThreadLocal<Random>(() =>
+        {
+            int seed;
+            lock (Global)
+            {
+                seed = Global.Next();
+            }
+
+            return new Random(seed);
+        });
+
+        /// <summary>
+        /// Replace the trace id of root activity.
+        /// </summary>
+        /// <param name="builder">Instance of <see cref="TracerProviderBuilder"/>.</param>
+        /// <returns>The instance of <see cref="TracerProviderBuilder"/>.</returns>
+        public static TracerProviderBuilder AddXRayActivityTraceIdGenerator(this TracerProviderBuilder builder)
+        {
+            var awsXRayActivityListener = new ActivityListener
+            {
+                ActivityStarted = (activity) =>
+                {
+                    // Replace every root activity's trace id with X-Ray compatiable trace id
+                    if (string.IsNullOrEmpty(activity.ParentId))
+                    {
+                        var awsXRayTraceId = GenerateAWSXRayCompatiableActivityTraceId();
+
+                        // Root node's parent id is no longer null, which will fail the sampler checker
+                        // 00-traceid-0000000000000000-00
+                        activity.SetParentId(awsXRayTraceId, default, activity.ActivityTraceFlags);
+                    }
+                },
+
+                ShouldListenTo = (_) => true,
+            };
+
+            ActivitySource.AddActivityListener(awsXRayActivityListener);
+
+            return builder;
+        }
+
+        public static ActivityTraceId GenerateAWSXRayCompatiableActivityTraceId()
+        {
+            var epoch = (int)DateTime.UtcNow.ToUnixTimeSeconds(); // first 8 digit as time stamp
+
+            var randomNumber = GenerateHexNumber(RandomNumberHexDigits); // remaining 24 random digit
+
+            return ActivityTraceId.CreateFromString(string.Concat(epoch.ToString("x", CultureInfo.InvariantCulture), randomNumber).AsSpan());
+        }
+
+        /// <summary>
+        /// Convert a given time to Unix time which is the number of seconds since 1st January 1970, 00:00:00 UTC.
+        /// </summary>
+        /// <param name="date">.Net representation of time.</param>
+        /// <returns>The number of seconds elapsed since 1970-01-01 00:00:00 UTC. The value is expressed in whole and fractional seconds with resolution of microsecond.</returns>
+        private static decimal ToUnixTimeSeconds(this DateTime date)
```

Comment:
Not sure if would prevent a duplication, today we have in OpenTelemetry/Internal/DateTimeOffsetExtensions.net452.cs a method similar to this. The only difference is that an extension for `DateTimeOffset` instead of `DateTime`